File: src/filetypes/filetype-lang.ts
import type { Language } from '~/filetypes/language.ts'
import type { Property } from '~/filetypes/property.ts'
import { Filetype } from '~/filetypes/filetype.ts'
import { Languages } from '~/filetypes/language.ts'
import { lineIterator } from '~/utils/line-iterator.ts'

const decoder = new TextDecoder()

export class FiletypeLang extends Filetype {
  #properties?: Record<string, Property>

  constructor(
    relpath: string,
    content: string,
    private readonly language: Language = Languages.en_US,
  ) {
    super(relpath, content)
  }

  get properties(): Record<string, Property> {
    if (!this.#properties) {
      this.#properties = this.parse()
    }
    return this.#properties
  }

  private parse(): Record<string, Property> {
    const properties: Record<string, Property> = {}
    const contentBytes = new TextEncoder().encode(this.content)
    for (const [, bytes, start, end] of lineIterator(contentBytes)) {
      if (bytes.length === 0 || bytes[0] === 0x23) { // 0x23 is '#'
        continue
      }

      // 1. Find the '=' byte (0x3D) without decoding the whole line
      const splitByteIndex = bytes.indexOf(0x3D)
      if (splitByteIndex === -1) {
        continue
      }

      // 2. Count code points up to the byte after '=' to get value start offset
      // This is extremely fast and avoids string slicing/UTF-16 complexity
      let valueCpOffset = 0
      for (let i = 0; i <= splitByteIndex; i++) {
        if ((bytes[i]! & 0xC0) !== 0x80) {
          valueCpOffset++
        }
      }

      const key = decoder.decode(bytes.subarray(0, splitByteIndex))
      const sKey = `lang|${key}`
      const value = decoder.decode(bytes.subarray(splitByteIndex + 1))
      const full = decoder.decode(bytes)

      properties[sKey] = {
        key: sKey,
        value,
        full,
        start: start + valueCpOffset,
        end,
      }
    }
    return properties
  }

  override getEnUsRelpath(): string {
    if (this.language === Languages.en_US) {
      return this.relpath
    }
    return this.relpath.replace(this.language, Languages.en_US)
  }

  override getTargetLanguageRelpath(targetLanguage: Language): string {
    if (this.language === targetLanguage) {
      return this.relpath
    }
    return this.relpath.replace(this.language, targetLanguage)
  }
}

File: src/gtnh_translation_compare/filetypes/filetype_lang.py
from typing import Dict

from gtnh_translation_compare.filetypes.filetype import Filetype
from gtnh_translation_compare.filetypes.language import Language
from gtnh_translation_compare.filetypes.property import Property
from gtnh_translation_compare.utils.line_iterator import line_iterator


class FiletypeLang(Filetype):
    def __init__(self, relpath: str, content: str, language: Language = Language.en_US):
        self._relpath = relpath
        self._content = content
        self._language = language

    def _get_relpath(self) -> str:
        return self._relpath

    def _get_content(self) -> str:
        return self._content

    def _get_properties(self, content: str) -> Dict[str, Property]:
        properties: Dict[str, Property] = {}
        for _, line, start, end in line_iterator(content):
            if line.startswith("#"):
                continue
            # noinspection DuplicatedCode
            split = line.split("=", 1)
            if len(split) != 2:
                continue
            key = split[0]
            s_key = f"lang|{key}"
            value = split[1]
            full = line
            properties[s_key] = Property(key=s_key, value=value, full=full, start=end - len(value), end=end)
        return properties

    def get_en_us_relpath(self) -> str:
        if self._language == Language.en_US:
            return self._relpath
        return self._relpath.replace(self._language.value, Language.en_US.value)

    def get_target_language_relpath(self, target_language: Language) -> str:
        if self._language == target_language:
            return self._relpath
        return self._relpath.replace(self._language.value, target_language.value)
